Objects/exceptions.c

```diff
@@ -1115,23 +1115,51 @@ OSError_str(PyOSErrorObject *self)
                                     self->winerror ? self->winerror: Py_None,
                                     self->strerror ? self->strerror: Py_None);
 #endif
+    PyObject *str = self->myerrno;
+    static PyObject *errorcodedict;
+    // Try to obtain the symbolic errno name (e.g., ENOENT).
+    if (str != NULL) {
+        if (errorcodedict == NULL) {
+            PyObject *errnomod = PyImport_ImportModule("errno");
+            if (errnomod == NULL) {
+                PyErr_Clear();
+                goto skip;
```

Comment:
I think this logic would be clearer by avoiding `goto`.

Probably a good way to do that is to pull this whole block into its own function.

Reply:
Using `goto` is fine, but the label should describe the target, such as `format_errno`. Just `skip` isn't meaningful.

Reply:
Hmm. I think most likely such a discussion would say not to use `goto` here. :slightly_smiling_face:

There's one very useful way of using `goto` in C which is done widely in the codebase: namely, to clean something up or tie up a loose end just before returning from a function. It serves much like a substitute for `finally:`.

I think in fact all the common label names you mention are examples of the names one might use in that use case. :wink:

Other cases where `goto` is worth it are quite rare, and I don't think this is one of them.

Reply:
Point taken. Jumping to a non-final label is rare in CPython's original C code. (One place I've seen it is in `tok_get` in Parser/tokenizer.c.) Jumping to the label in this case didn't bother me because it's a single step forward, as opposed to spaghetti that twists my brain into a knot. That said, adding a function to get the error name would improve readability.

Reply:
That is in fact exactly the example that came to my mind too. :-) It does contribute to making that logic rather tricky to understand; but of course the inner loop of the tokenizer is quite a hot path, where saving a few cycles potentially justifies some virtuoso loop-optimization tricks. When I looked hard at that code a couple of years ago, I remember thinking that it was one of the rare cases where a non-final goto makes sense.

Even there, who knows; computers have changed a lot since Guido wrote that code, in 1990 or so. (In the early history, the loop looks pretty much like it does now.) It's possible that something more straightforward might be preferable today.

Reply:
I will use `goto` in `OSError_str()` unless a core dev says not to.